### DISCUSSION TITLE
tutorials/3d/procedural_geometry/arraymesh

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#208] [User-contributed notes]
# tutorials/3d/procedural_geometry/arraymesh

This tutorial will present the basics of using an ArrayMesh. To do so, we will use the function add_surface_from_arrays(), which takes up to five parameters. The first two are required, while the l...

https://docs.godotengine.org/en/stable/tutorials/3d/procedural_geometry/arraymesh.html

**Comment by dr-jam** [comment]
The GDScript examples contain errors. In particular, the following example wrongly associates `add_surface_from_arrays()` with the `MeshInstance3D` class. It is a function in the `ArrayMesh`class.
```gdscript
surface_array[Mesh.ARRAY_VERTEX] = verts
surface_array[Mesh.ARRAY_TEX_UV] = uvs
surface_array[Mesh.ARRAY_NORMAL] = normals
surface_array[Mesh.ARRAY_INDEX] = indices

# No blendshapes, lods, or compression used.
mesh.add_surface_from_arrays(Mesh.PRIMITIVE_TRIANGLES, surface_array)
```

**Reply by Calinou** [reply]
The `mesh` identifier refers to the ArrayMesh, not a MeshInstance3D. If you attach a script to a MeshInstance3D, you access its mesh data by reading `mesh` directly.

See the text at the top:

> In the editor, create a MeshInstance3D and add an ArrayMesh to it in the Inspector. Normally, adding an ArrayMesh in the editor is not useful, but in this case it allows us to access the ArrayMesh from code without creating one.

**Reply by dr-jam** [reply]
Thanks! I must've gotten lost in the sauce.

**Comment by Arikel135** [comment]
This can be conbined with fast noise and create procedural terrain.

**Comment by BieniekAlexander** [comment]
Could somebody add more conceptual details to this page, or references to such content? It's assumed that the reader knows the meaning of  `ARRAY_VERTEX`, `ARRAY_INDEX`, `ARRAY_TEX_UV`, etc., and I'm having trouble with learning these details.

**Comment by NecroDes** [comment]
How can I make a flat shading like when using SurfaceTool.add_smooth_group(-1)

**Reply by NecroDes** [reply]
I found a solution. You need to use a normalized cross product between two edges.

**Comment by AMAltundag** [comment]
I spent a fair bit messing around with this thing, so here are some notes of mine. Its a bit messy, but I suppose it might be a useful reference for people messing around with it.

My experiments were with primitive triangles, utilzing all the arrays including the index array.

* ARRAY_VERTEX = this is just a single point, you use three of these for lighting
* ARRAY_NORMAL = normals are used for many things, face culling, lighting etc, it is a vector that points into a direction
* ARRAY_TANGENT = no clue, never had to mess with it.
* ARRAY_COLOR = this allows you to store a color, this is per vertex, your shader can access these
* ARRAY_TEX_UV = UV coordinates are points on your texture, there is an order to them, you get one per vertex, three vertices make a triangle, the three vertices stretch to become the texture
* ARRAY_TEX_UV2 = godot can handle multiple uvs, but i am unsure if they are used within godot. I heard talks about how it can be used for lighting, didnt have to mess with this
* ARRAY_BONES = despite the docs saying 4, it can actually be 8 depending on the flags. This is relevant for animation rigs, didnt have to mess with it. I assume this just lists up which bones can affect it.
* ARRAY_WEIGHTS = can 4 or 8, i assume this is a multiplier to how the bone deformation affects the vertice
* ARRAY_INDEX = three indices make one triangle, there might (?) be an order to these. This is just a number that points to a vertex.

Outside of the index array, the other arrays are ordered and connected. 1 vertex, 1 normal, 4 tangent, 1 color, 1 uv, 1 uv2, 4/8 bones, 4/8 weights.

One notable detail, most 3D software will optimize your mesh before exporting. In certain circumstances, you do not need the same vertex multiple times. If all values overlap, you can just have one vertex and refer it multiple times via the index array. You may have to normalize the mesh yourself, pulling all the vertices apart if you wish to manipulate parts of the thing otherwise you may have unintended consequences.

Not quite obvious from this page, and I didn't delve deeper into it, but arrays 6 7 8 9 can get custom colors on them too, I assume these are accessible via shaders and are mapped per vertex. Using these four colors (I suppose you could have a fifth using the vertex color?) you can encode the RGBA components into it (or just RGB? didnt try, not sure if it can handle RGBA) you can encode 4x3 (or 4x4) numbers ranging 0-255 into the array. Meaning you might be able to use the custom colors to blend between 12/16 textures - which should be enough for a terrain mesh (especially if you multiply the vertex color with the texture, saving you from wasting textures on colored variants.)

**Comment by emanvidmaker** [comment]
For multiple material support, make a new surface_array for each material with the geometry that will have it. call mesh.add_surface_from_arrays() with the new surface_array and a new surface/material slot will be available now

**Comment by bradgian** [comment]
In this tutorial, it is not clear that "mesh" is an ArrayMesh or how to create it. I tried to add it as a node using the editor, but "ArrayMesh" is not known. I eventually looked up the ArrayMesh doc, which stated:
```
      var arr_mesh = ArrayMesh.new()
```
So then I added this line at the top of the ready function:
```
     mesh = ArrayMesh.new()
```
I'm brand new to godot so not sure if this is correct, but at least I was able to run the scene without crashing. Can someone verify this for me?

**Reply by Calinou** [reply]
ArrayMesh is not a *node*, but a *resource*. This is why it doesn't appear in the Create New Node dialog. It will appear in the Create New Resource dialog, but there is no dedicated visual editor for it anyway, so you should create it from code instead.

> I'm brand new to godot so not sure if this is correct

Yes, this is the correct way to do it.
